tests/components/monzo/test_config_flow.py

```diff
@@ -136,3 +140,153 @@ async def test_config_non_unique_profile(
+    aioclient_mock.post(
+        OAUTH2_TOKEN,
+        json={
+            "refresh_token": "mock-refresh-token",
+            "access_token": "new-mock-access-token",
+            "type": "Bearer",
+            "expires_in": 60,
+            "user_id": str(USER_ID),
+        },
+    )
+
+    result = await hass.config_entries.flow.async_configure(result["flow_id"])
+
+    assert result["type"] is FlowResultType.FORM
+    assert result["step_id"] == "await_approval_confirmation"
+    assert polling_config_entry.data["token"]["access_token"] == "mock-access-token"
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"], user_input={"confirm": True}
+    )
+
+    assert len(hass.config_entries.async_entries(DOMAIN)) == 1
+    assert result
+    assert result["type"] is FlowResultType.ABORT
+    assert result["reason"] == "reauth_successful"
+    assert polling_config_entry.data["token"]["access_token"] == "new-mock-access-token"
+
+
+async def test_config_reauth_wrong_account(
+    hass: HomeAssistant,
+    hass_client_no_auth: ClientSessionGenerator,
+    aioclient_mock: AiohttpClientMocker,
+    polling_config_entry: MockConfigEntry,
+    current_request_with_host: None,
+) -> None:
+    """Test reauth with wrong account."""
+    await setup_integration(hass, polling_config_entry)
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN,
+        context={
+            "source": SOURCE_REAUTH,
+            "entry_id": polling_config_entry.entry_id,
+        },
+        data=polling_config_entry.data,
+    )
+    assert result["type"] is FlowResultType.FORM
+    assert result["step_id"] == "reauth_confirm"
+
+    result = await hass.config_entries.flow.async_configure(result["flow_id"], {})
+    state = config_entry_oauth2_flow._encode_jwt(
+        hass,
+        {
+            "flow_id": result["flow_id"],
+            "redirect_uri": "https://example.com/auth/external/callback",
+        },
+    )
+    assert result["url"] == (
+        f"{OAUTH2_AUTHORIZE}/?"
+        f"response_type=code&client_id={CLIENT_ID}&"
+        "redirect_uri=https://example.com/auth/external/callback&"
+        f"state={state}"
+    )
+    client = await hass_client_no_auth()
+    resp = await client.get(f"/auth/external/callback?code=abcd&state={state}")
+    assert resp.status == 200
+    assert resp.headers["content-type"] == "text/html; charset=utf-8"
+
+    aioclient_mock.clear_requests()
+    aioclient_mock.post(
+        OAUTH2_TOKEN,
+        json={
+            "refresh_token": "mock-refresh-token",
+            "access_token": "mock-access-token",
+            "type": "Bearer",
+            "expires_in": 60,
+            "user_id": 12346,
+        },
+    )
+
+    result = await hass.config_entries.flow.async_configure(result["flow_id"])
+    assert result
+    assert result["type"] == FlowResultType.ABORT
+    assert result["reason"] == "wrong_account"
+
+
+async def test_api_can_trigger_reauth(
```

Comment:
I believe the common pattern is to put a test like this in `test_init.py`

Reply:
Is there a good reason for this? It seems odd to put it in `test_init.py` since it doesn't actually test anything in `__init__.py` and all the other reauth tests are in `test_config_flow.py`. I could maybe understand `test_coordinator.py` given that's what raises the error , but even that's unintuitive

Reply:
My impression is that this is exercising the code here https://github.com/home-assistant/core/blob/e4130480c3350f4ad1ead87b805ace61f319cd88/homeassistant/components/monzo/__init__.py#L31 -- though i see now its testing a later update.  Typically the code is exercising the startup behavior. I suspect if you moved this before the `setup_integration` it would test the startup behavior.

Reply:
I see, still unusual though since the test is more about any refresh triggering reauth rather than init behaviour. Anyway, in my case I suppose it's definitely in the correct place as is

Reply:
My impression is that the rationale for this pattern is that typically the test would exercise exceptions throw during first init. Initi would be a common place where reauth might fail for the first time. Essentially, the idea is that it is testing the call to `async_config_entry_first_refresh` fails and is handled properly by the home assistant core code end to end.

For example in wiithings the tests are here:
https://github.com/home-assistant/core/blob/e4130480c3350f4ad1ead87b805ace61f319cd88/tests/components/withings/test_init.py#L144

Which are exercising this code in init:
https://github.com/home-assistant/core/blob/e4130480c3350f4ad1ead87b805ace61f319cd88/homeassistant/components/withings/__init__.py#L128

Perhaps part of this pattern is historical, established before update coordinators also had baked in logic for handling these exceptions and it was common for the code in init to actually initialize reauth itself e.g. https://github.com/home-assistant/core/pull/48962 -- but i think it still makes sense from the perspective of exercising the code on init/startup as the primary thing to exercise.

Reply:
Again, I see what you mean, but I think in typical use auth is much more likely to expire while HA is running rather than when stopped - eg. my instance has 99.87% uptime. More importantly, this test shouldn't be first refresh _or_ later refresh specific; we want to know that reauth is triggered regardless of when. It might even be better suited to be in `test_coordinator.py`, though keeping it with the rest of the reauth tests is probably a better fit for now